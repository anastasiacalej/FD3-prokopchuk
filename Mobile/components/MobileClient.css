

span.MobileClientBalance
{
  float: right;
  width: 30px;
  font-weight: bold;
  text-align: right;
}

span.MobileClientFIO
{
  margin-right: 30px;
}

.StatusActive
{
  background-color:green;
  text-align: center;
}

.StatusBlocked
{
  background-color:red;
  text-align: center;
}

.Edit
{
  height: 20px;
  font-style: italic;
  background-color: #d9d7d797;
}

.Highlight{
  background-color:#299e2d91;
}

 input.Validate::placeholder{
   color:red;
 }
