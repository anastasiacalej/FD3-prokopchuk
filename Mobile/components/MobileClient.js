import React from 'react';
import PropTypes from 'prop-types';

import './MobileClient.css';
import {eventsClicked} from './events';

class MobileClient extends React.PureComponent {

  static propTypes = {
    info:PropTypes.shape({
      id: PropTypes.number.isRequired,
      f: PropTypes.string,
      i: PropTypes.string,
      o: PropTypes.string,
      balance: PropTypes.number,
    }),
  };

  inputRefFam=null;//ссылка на нужный input
  inputRefIm=null;//ссылка на нужный input
  inputRefOtch=null;//ссылка на нужный input
  inputRefBalance=null;//ссылка на нужный input
  editButton=null;
 
  //узнаем ссылки на нужные inputы
  setRefFam = (ref)=>{
    this.inputRefFam=ref;
  }

  setRefIm = (ref)=>{
    this.inputRefIm=ref;
  }

  setRefOtch = (ref)=>{
    this.inputRefOtch=ref;
  }

  setRefBalance = (ref)=>{
    this.inputRefBalance=ref;
  }

  setEditButton = (ref)=>{
    this.editButton=ref;
  }

  validate = ()=>{
   if (  this.inputRefFam.value==='') 
   {this.inputRefFam.placeholder='Введите фамилию';this.inputRefFam.className='Edit Validate' }
   if (  this.inputRefIm.value==='') 
   {this.inputRefIm.placeholder='Введите имя';this.inputRefIm.className='Edit Validate'}
   if (  this.inputRefOtch.value==='') 
   {this.inputRefOtch.placeholder='Введите отчество';this.inputRefOtch.className='Edit Validate'}
   if (  isNaN(parseInt(this.inputRefBalance.value))) 
   {this.inputRefBalance.value=''; this.inputRefBalance.placeholder='Введите цифру';this.inputRefBalance.className='Edit Validate'}
  }

  saveEditClient = ()=>{
   
    if (this.inputRefFam.value!==''
      &&this.inputRefIm.value!==''
      &&this.inputRefOtch.value!==''
      &&!isNaN(parseInt(this.inputRefBalance.value))){
    //формируем новую информацию о клиенте в хэш:    
    let newInfo= {...this.props.info,f:this.inputRefFam.value, i:this.inputRefIm.value, 
      o:this.inputRefOtch.value, balance:parseInt(this.inputRefBalance.value)};

    eventsClicked.emit('ESaveClientClick', newInfo);
    }else{
      this.validate();
    }
  }

  saveAddClient = ()=>{ 
    if (this.inputRefFam.value!==''
      &&this.inputRefIm.value!==''
      &&this.inputRefOtch.value!==''
      &&!isNaN(parseInt(this.inputRefBalance.value))){
    //формируем полученную информацию о новом клиенте в хэш:   
    let newInfo= {...this.props.info,f:this.inputRefFam.value, i:this.inputRefIm.value, 
      o:this.inputRefOtch.value, balance:parseInt(this.inputRefBalance.value)}; 

    eventsClicked.emit('ESaveAddClientClick', newInfo);
  }else{
    this.validate();
  }
  }

  deleteClient = () =>{
    eventsClicked.emit('EDeleteClientClick', this.props.info.id);
  }

  editClient = () =>{
    eventsClicked.emit('EEditClientClick', this.props.info.id);
  }

  cancelEditClient = ()=>{
    eventsClicked.emit('EEditCancelClientClick');
  }

  cancelAddClient =()=>{
    eventsClicked.emit('EAddCancelClientClick');
  }

  disableEdit = ()=>{
    if(this.editButton)
    this.editButton.disabled=true;
  }

  ableEdit =()=>{
    if(this.editButton)
    this.editButton.disabled=false;
  }

  componentDidMount = () =>{
    eventsClicked.addListener('EAddCurrentClientClick',this.disableEdit);
    eventsClicked.addListener('EAddedClientClick',this.ableEdit);
  };

  componentWillUnmount = () =>{
    eventsClicked.removeListener('EAddCurrentClientClick',this.disableEdit);
    eventsClicked.removeListener('EAddedClientClick',this.ableEdit);
  };

  render() {

    console.log("MobileClient id="+this.props.info.id+" render");
    
    if (this.props.editClient!==this.props.info.id&&!this.props.addClient)
    return (
      <tr className='MobileClient'>        
        <td>{this.props.info.f}</td>
        <td>{this.props.info.i}</td>
        <td>{this.props.info.o}</td>
        <td>{this.props.info.balance}</td>
        <td className={this.props.info.balance>=0?'StatusActive':'StatusBlocked'}>
          {this.props.info.balance>=0?'active':'blocked'}</td>
        <td>
          <input className='ButtonName' type="button" value="Редактировать" onClick={this.editClient} ref={this.setEditButton}/>
        </td>
        <td>
          <input className='ButtonName' type="button" value="Удалить" onClick={this.deleteClient} />
        </td>
      </tr>
    );

    if (this.props.editClient===this.props.info.id&&!this.props.addClient)
    return (
      <tr className='MobileClient' className='Highlight'>        
        <td>
          <input className='Edit' type="text" defaultValue={this.props.info.f} ref={this.setRefFam} />
        </td>
        <td>
          <input className='Edit' type="text" defaultValue={this.props.info.i} ref={this.setRefIm} />
        </td>
        <td>
          <input className='Edit' type="text" defaultValue={this.props.info.o} ref={this.setRefOtch} />
        </td>
        <td>
          <input className='Edit' type="text" defaultValue={this.props.info.balance} ref={this.setRefBalance} />
        </td>
        <td className={this.props.info.balance>=0?'StatusActive':'StatusBlocked'}>
          {this.props.info.balance>=0?'active':'blocked'}</td>
        <td>
          <input className='ButtonName' type="button" value="Cохранить" onClick={this.saveEditClient} />
        </td>
        <td>
          <input className='ButtonName' type="button" value="Отмена" onClick={this.cancelEditClient} />
        </td>
      </tr>
    );

    if (this.props.addClient)
    return (
      <tr className='MobileClient' className='Highlight'>        
        <td>
          <input className='Edit' type="text" placeholder="Введите фамилию" ref={this.setRefFam} />
        </td>
        <td>
          <input className='Edit' type="text" placeholder="Введите имя" ref={this.setRefIm} />
        </td>
        <td>
          <input className='Edit' type="text" placeholder="Введите отчество" ref={this.setRefOtch} />
        </td>
        <td>
          <input className='Edit' type="text" placeholder="Введите баланс" ref={this.setRefBalance} />
        </td>
        <td></td>
        <td>
          <input className='ButtonName' type="button" value="Добавить" onClick={this.saveAddClient} />
        </td>
        <td>
          <input className='ButtonName' type="button" value="Отмена" onClick={this.cancelAddClient} />
        </td>
      </tr>
    );

  }

}

export default MobileClient;
