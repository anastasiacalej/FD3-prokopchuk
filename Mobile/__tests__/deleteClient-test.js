"use strict";

import React from 'react';
import renderer from 'react-test-renderer';

import MobileClient from '../components/MobileClient';

let clientsArr=[ 
  {id:101, f:"Иванов", i:"Иван", o:"Иванович", balance:200}, 
  {id:102, f:"Сидоров",i:"Сидор", o:"Сидорович", balance:250}, 
  {id:103, f:"Петров", i:"Петр", o:"Петрович", balance:180},
  {id:104, f:"Григорьев", i:"Григорий", o:"Григорьевич", balance:-220},
];

test('работа кнопки "Удалить"', () => {

  // создаём тестовую версию компонента
  const component = renderer.create(
    <MobileClient key={clientsArr[0].id} info={clientsArr[0]} //передать можно любого клиента
    //если id очередного клиента равно id редактируемого клиента - передадим этот id в пропсах (для кнопок 'сохранить' и 'отмена')
    //иначе ложное значение - для проверки кнопки 'Удалить':
    editClient={null}
    //всегда передаем null: 
    addClient={null}/>
  );

  // получаем снэпшот (HTML-снимок) компонента для сверки, что вёрстка не испортилась
  let componentTree=component.toJSON();
  expect(componentTree).toMatchSnapshot();

  // найдём в вёрстке компонента саму кнопку "Удалить"
  const buttonElem = component.root.find(el => el.type=='input' && el.props.value == "Удалить"); 
  
  // и "нажмём" на неё
  buttonElem.props.onClick();

  // получаем уже изменённый снэпшот
  componentTree=component.toJSON();
  expect(componentTree).toMatchSnapshot();

  })
  