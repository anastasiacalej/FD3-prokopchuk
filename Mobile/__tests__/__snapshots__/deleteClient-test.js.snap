// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`работа кнопки "Удалить" 1`] = `
<tr
  className="MobileClient"
>
  <td>
    Иванов
  </td>
  <td>
    Иван
  </td>
  <td>
    Иванович
  </td>
  <td>
    200
  </td>
  <td
    className="StatusActive"
  >
    active
  </td>
  <td>
    <input
      className="ButtonName"
      onClick={[Function]}
      type="button"
      value="Редактировать"
    />
  </td>
  <td>
    <input
      className="ButtonName"
      onClick={[Function]}
      type="button"
      value="Удалить"
    />
  </td>
</tr>
`;

exports[`работа кнопки "Удалить" 2`] = `
<tr
  className="MobileClient"
>
  <td>
    Иванов
  </td>
  <td>
    Иван
  </td>
  <td>
    Иванович
  </td>
  <td>
    200
  </td>
  <td
    className="StatusActive"
  >
    active
  </td>
  <td>
    <input
      className="ButtonName"
      onClick={[Function]}
      type="button"
      value="Редактировать"
    />
  </td>
  <td>
    <input
      className="ButtonName"
      onClick={[Function]}
      type="button"
      value="Удалить"
    />
  </td>
</tr>
`;
